temporian/utils/rtcheck.py

```diff
@@ -0,0 +1,340 @@
+
+    Args:
+        fn: Function to instrument.
+
+    Returns:
+        Instrumented function.
+    """
+    is_compiled = False
+
+    if hasattr(fn, "__wrapped__"):
+        signature_fn = fn.__wrapped__
+        if hasattr(signature_fn, "is_tp_compiled"):
+            is_compiled = getattr(signature_fn, "is_tp_compiled")
+    else:
+        signature_fn = fn
+    signature = inspect.signature(signature_fn)
+
+    @wraps(fn)
+    def wrapper(*args, **kwargs):
+        trace = _Trace().add_context(f'When checking function "{fn.__name__}".')
+
+        try:
+            # Check inputs
+            all_args = signature.bind(*args, **kwargs)
+            for arg_key, arg_value in all_args.arguments.items():
+                if arg_key not in signature.parameters:
+                    raise ValueError(f'Unexpected argument "{arg_key}"')
+                param = signature.parameters[arg_key]
+
+                if param.kind in [
+                    inspect.Parameter.POSITIONAL_ONLY,
+                    inspect.Parameter.POSITIONAL_OR_KEYWORD,
+                ]:
+                    _check_annotation(
+                        trace.add_context(
+                            f'When checking argument "{arg_key}".'
+                        ),
+                        is_compiled,
+                        arg_value,
+                        param.annotation,
+                    )
+                elif param.kind is inspect.Parameter.VAR_POSITIONAL:
+                    _check_annotation_list_or_set_or_uniform_tuple(
+                        trace.add_context(
+                            f'When checking argument "{arg_key}".'
+                        ),
+                        is_compiled,
+                        arg_value,
+                        [param.annotation],
+                    )
+                elif param.kind is inspect.Parameter.VAR_KEYWORD:
+                    for sub_key, sub_value in arg_value.items():
+                        _check_annotation(
+                            trace.add_context(
+                                f'When checking key "{sub_key}" of argument'
+                                f' "{arg_key}".'
+                            ),
+                            is_compiled,
+                            sub_value,
+                            param.annotation,
+                        )
+        except ValueError as e:
+            # Reset the stack trace of the exception.
+            e.__traceback__ = None
+            raise e
+
+        output = fn(*args, **kwargs)
+
+        try:
+            # Check outputs
+            _check_annotation(
+                trace.add_context("When checking returned value."),
+                is_compiled,
+                output,
+                signature.return_annotation,
+            )
+        except ValueError as e:
+            # Reset the stack trace of the exception.
+            e.__traceback__ = None
+            raise e
```

Comment:
The main benefit of the rtcheck is to prevent error of user in colab. So, I think an error is good by default. However, having a way to disable the check (e.g., in case of an internal bug) is also good.
*Adding instruction on how to disable the check in the error message.